src/cryptography/hazmat/backends/openssl/backend.py

```diff
@@ -171,6 +171,15 @@ def _encode_subject_alt_name(backend, san):
             )
             assert obj != backend._ffi.NULL
             gn.d.registeredID = obj
+        elif isinstance(alt_name, x509.DirectoryName):
+            gn = backend._lib.GENERAL_NAME_new()
+            assert gn != backend._ffi.NULL
+            name = _encode_name(backend, alt_name.value)
+            # _encode_name registers the X509_NAME for gc so we'll duplicate
+            # a new one that is not gc'd for the struct
+            name = backend._lib.X509_NAME_dup(name)
```

Comment:
Not wild about a `gc` param; I'd prefer two functions (obviously one which
calls the other)

On Sat, Jul 11, 2015 at 12:29 AM, Paul Kehrer <notifications@github.com>
wrote:

> In src/cryptography/hazmat/backends/openssl/backend.py
> <https://github.com/pyca/cryptography/pull/2134#discussion_r34409541>:
>
> > @@ -171,6 +171,15 @@ def _encode_subject_alt_name(backend, san):
> >              )
> >              assert obj != backend._ffi.NULL
> >              gn.d.registeredID = obj
> > +        elif isinstance(alt_name, x509.DirectoryName):
> > +            gn = backend._lib.GENERAL_NAME_new()
> > +            assert gn != backend._ffi.NULL
> > +            name = _encode_name(backend, alt_name.value)
> > +            # _encode_name registers the X509_NAME for gc so we'll duplicate
> > +            # a new one that is not gc'd for the struct
> > +            name = backend._lib.X509_NAME_dup(name)
>
> What about changing _encode_name to take a gc boolean arg? No default, so
> current call sites would become _encode_name(backend, <data>, gc=True)
>
> Whatever pattern we choose will be duplicated for _encode_asn1_str and
> maybe _txt2obj as well.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2134/files#r34409541>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084